### ISSUE TITLE
Analytics is enabled by default in OSS version

### ISSUE THREAD

**Issue by markkolich** [liquibase/liquibase#6503]
### Search first

- [x] I searched and no similar issues were found

### Description

I upgraded a personal project from OSS Liquibase version `4.26.0` to version `4.30.0` (the latest as of this writing) and started seeing this exception in my application logs at runtime:

```
Exception in thread "Thread-2" java.lang.RuntimeException: java.net.ConnectException: Connection refused
	at liquibase.analytics.LiquibaseAnalyticsListener.lambda$handleEvent$2(LiquibaseAnalyticsListener.java:110)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.Net.connect0(Native Method)
	....
	at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.getOutputStream(HttpsURLConnectionImpl.java:219)
	at liquibase.analytics.LiquibaseAnalyticsListener.lambda$handleEvent$2(LiquibaseAnalyticsListener.java:101)
	... 1 more
```

The exception above is a result of my [Pi-hole](https://pi-hole.net/) DNS blocking traffic to host `api.segment.io`, a known analytics and ad tracking host.

Debugging, it appears that OSS Liquibase `4.30.0` is now, _by default_, phoning home and attempting to post usage data to `https://api.segment.io/v1/batch` that contains information about my Liquibase usage. Looks like LB is attempting to POST a JSON blob that captures my database, database version, OS platform, Java version, changeset count, command, etc.

I understand that this was added here https://github.com/liquibase/liquibase/pull/6412, and I'm using the OSS version ... so I get what I get.

And, I know it can disabled by setting `-Dliquibase.analytics.enabled=false` (which works):

```java
// https://github.com/liquibase/liquibase/issues/6501#issuecomment-2464909449
System.setProperty("liquibase.analytics.enabled", "false");
```

However, it feels odd and disingenuous to have a library on my classpath phoning home analytics data without my consent. I would have never noticed this unless requests to `api.segment.io` started failing, and I suspect the vast majority of your other OSS users are probably in the same boat.

I might have missed it, but I don't recall seeing any mention in your OSS license or elsewhere that states using the OSS version implies consent for LB to collect and share usage data. And to be honest, I don't know of any other OSS project that does this.

I understand you made an explicit choice to enable this behavior, but as a privacy centric developer, I feel strongly that this kind of analytics and usage reporting functionality should be opt-in and _disabled/off by default_.

Thanks for listening!

CC // tagging a few folks from https://github.com/liquibase/liquibase/issues/6501 @rdhzl @lbadam @freddiN @filipelautert

### Steps To Reproduce

Upgrade to version `4.30.0` and see unexpected outbound network traffic to `api.segment.io`.

### Expected/Desired Behavior

IMHO, `liquibase.analytics.enabled=false` should be the default behavior. For those in the OSS community who wish to give back and are OK sharing such data, we can opt-in manually on a case-by-case basis.

### Liquibase Version

4.30.0

### Additional Context

From my debugger, here's a sample JSON payload that Liquibase is attempting to POST to `api.segment.io`:

```json
{
  "batch": [
    {
      "anonymousId": "18a95bec-6598-3e91-acb2-1b58eacbeef2",
      "context": "null",
      "event": "liquibase-command-executed",
      "messageId": "c967850b-f334-4ed7-ba60-819b1044ad1a",
      "properties": {
        "command": "validate",
        "reportsEnabled": false,
        "dbclhEnabled": false,
        "structuredLogsEnabled": false,
        "operationOutcome": "success",
        "exceptionClass": "null",
        "chlog_sql": 0,
        "chlog_formattedSql": 0,
        "chlog_xml": 4,
        "chlog_json": 0,
        "chlog_yaml": 0,
        "databasePlatform": "...redacted...",
        "databaseVersion": "...redacted...",
        "liquibaseVersion": "4.30.0",
        "liquibaseInterface": "JavaAPI",
        "javaVersion": "...redacted...",
        "os": "...redacted...",
        "osVersion": "...redacted...",
        "osArch": "...redacted...",
        "isDocker": false,
        "isLiquibaseDocker": false,
        "isAwsLiquibaseDocker": false,
        "isGithubActions": false,
        "isCi": false,
        "ext_mongoDb_commercial": "null",
        "ext_dynamoDb_commercial": "null",
        "ext_checks": "null",
        "ext_awsSecrets": "null",
        "ext_awsS3": "null",
        "ext_hashicorpVault": "null",
        "ext_bigQuery_commercial": "null",
        "ext_databricks_commercial": "null",
        "ext_bigQueryOss": "null",
        "ext_mongoDbOss": "null",
        "ext_awsLicenseService": "null"
      },
      "type": "track",
      "userId": "null"
    }
  ],
  "context": "null",
  "writeKey": "jtKzsxcg4SKsxRRJEre0dc0kxkk4umqt"
}
```

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR (Thank you!)

**Comment by rdhzl** [comment]
I appreciate your filing this, if for nothing else than it makes it easy to see what telemetry is being shipped back.

In my case I've got to support Liquibase running in a variety of ways (command line, Gradle, programmatically in Java mains) in multiple environments (developer envs up through production). So there's seemingly no singular way for me to shut it off that's once and done, and I now have to now explain this to my leadership. =/

I understand the desire to better understand how the software is being used, and I appreciate what the OSS version provides the community, but maybe just send me a survey instead.

**Comment by mariochampion** [comment]
hey @markkolich  thanks for the feedback, In case you havent seen the information at https://docs.liquibase.com/analytics 

Opt out
If you do not want to participate in Liquibase anonymous analytics, you can opt out at any time using the --[analytics-enabled](https://docs.liquibase.com/parameters/analytics-enabled.html) parameter.

**Opt out per CLI command:** specify the parameter --analytics-enabled=false. This turns off analytics gathering for that single occurrence of a command. For example:
liquibase --analytics-enabled=false status

**Opt out globally:** add liquibase.analytics.enabled: false to your [defaults file](https://docs.liquibase.com/concepts/connections/creating-config-properties.html) or use the [environment variable](https://docs.liquibase.com/concepts/connections/liquibase-environment-variables.html) LIQUIBASE_ANALYTICS_ENABLED=false.

hope this helps!

**Comment by lbadam** [comment]
@markkolich thank for taking the time to write this issue up. I (and all of at Liquibase) definitely understand the concern around privacy, and the desire to have this be opt-out by default.

We designed this similarly to [Homebrew's Anonymous Analytics](https://docs.brew.sh/Analytics) making sure that:

- there is no PII or way to track individual users
- there are multiple ways to opt-out prior to analytics data being sent out (as @mariochampion has outlined above)

Throughout this process, we have been upfront with these changes. We talked about this coming in v4.30.0 (#6412). We published it in [release notes](https://github.com/liquibase/liquibase/releases/tag/v4.30.0) with a note about turning it off before data was sent, and we have even published a [post-launch blog](https://www.liquibase.com/blog/product-update-liquibase-now-collects-anonymous-usage-analytics) for users to read. We also updated distribution licenses in #6411 to reflect these changes. 

We believe collecting usage data in this way will help us understand how the millions of users Liquibase has are actually using the product so we can make it even better, but we also believe that it preserves the privacy of our users which is extremely important to us.

**Comment by markkolich** [comment]
Thanks for the followups @mariochampion & @lbadam.

As mentioned in my original post, I know you can disable this and opt-out.

However, my issue is that this is enabled by default which I think is a serious erosion of trust.

This move makes me wonder: if you’re comfortable putting something like this into a JAR on my classpath, what else might be captured without my consent? It makes me very wary of the potential to capture more potentially sensitive information — like database names, tables, and schema details.

I understand that you have your reasons, but as a consumer of OSS, I do not feel that this style of analytics and metrics data gathering should be enabled by default. I contend it should be the exact opposite: this should be _disabled by default_, with an option to let users explicitly opt-in if they choose to share how they are using your library.

And to @rdhzl's point, a survey would probably have been a fine alternative.

**Comment by StevenMassaro** [comment]
@brandonlenz Your property is correct. Refer to the help output if you're ever unsure.

```
      --analytics-enabled=PARAM
                             Enable or disable sending product usage data and
                               analytics to Liquibase. Learn more at https:
                               //docs.liquibase.com/analytics. DEFAULT: true
                               for OSS users | false for PRO users
                             (defaults file: 'liquibase.analytics.enabled',
                               environment variable:
                               'LIQUIBASE_ANALYTICS_ENABLED')
```

There is another issue for Spring: https://github.com/liquibase/liquibase/issues/6501

**Comment by brandonlenz** [comment]
> [@brandonlenz](https://github.com/brandonlenz) Your property is correct. Refer to the help output if you're ever unsure.
> 
> ```
>       --analytics-enabled=PARAM
>                              Enable or disable sending product usage data and
>                                analytics to Liquibase. Learn more at https:
>                                //docs.liquibase.com/analytics. DEFAULT: true
>                                for OSS users | false for PRO users
>                              (defaults file: 'liquibase.analytics.enabled',
>                                environment variable:
>                                'LIQUIBASE_ANALYTICS_ENABLED')
> ```
> 
> There is another issue for Spring: [#6501](https://github.com/liquibase/liquibase/issues/6501)

Thank you very much!

**Comment by t-beckmann** [comment]
From a data protection perspective not having this feature as an opt-in is a serious breach of trust.

**Comment by brandonlenz** [comment]
Despite my notes above asking about how to disable for spring boot apps, I'm going to also comment to explicitly note that I am in agreement with the rest of the community that analytics should be off by default.

I get it "How else will we get OSS users to opt in?". I don't have an answer for that, but that sounds like an interesting problem to solve and incentive structures to think about. Or just document why analytics would be helpful more front-and-center in the Liquibase docs and see if you can get folks who do not have strict security requirements on their projects to join in.

If anyone is up for a PR, looks like this is the file to change: https://github.com/liquibase/liquibase/blob/3b735c4f188c2fb2105d3b664b1eeecfa03d305f/liquibase-standard/src/main/java/liquibase/analytics/configuration/AnalyticsArgs.java#L83

**Comment by erik-meuwese-topicus** [comment]
[sideeffffect](https://github.com/sideeffffect) wrote in the last comment of the pull request https://github.com/liquibase/liquibase/pull/6412 a link to an easy solution to turn it off programmatically: https://github.com/sideeffffect/liquibase-doobie/pull/426/files 

```
Scope.enter(Map.of(AnalyticsArgs.ENABLED.getKey(), Boolean.FALSE));
```

**Comment by Haarolean** [comment]
Scanning an app's classpath for jars phoning home was certainly not on my 2024 agenda.

> Throughout this process, we have been upfront with these changes. We talked about this coming in v4.30.0 ([#6412](https://github.com/liquibase/liquibase/pull/6412)). We published it in [release notes](https://github.com/liquibase/liquibase/releases/tag/v4.30.0) with a note about turning it off before data was sent, and we have even published a [post-launch blog](https://www.liquibase.com/blog/product-update-liquibase-now-collects-anonymous-usage-analytics) for users to read. We also updated distribution licenses in [#6411](https://github.com/liquibase/liquibase/pull/6411) to reflect these changes.

Regardless of how many blog posts are written about this, it doesn't help much. Java apps often include tens of direct dependencies and dozens of transitive ones, and I'm not willing to spend my leisure time poring over every blog post to avoid missing a stealthy analytics opt-in announcement.

> From a data protection perspective not having this feature as an opt-in is a serious breach of trust.

Precisely. Introducing such a feature enabled by default erodes trust and creates a slippery slope. 
As an OSS maintainer, I'm well aware of the flood of questions users raise about every outgoing third-party call -- even something as innocuous as a version check. Considering this, I'm certain the lack of backlash here is simply because most people are unaware that their apps' dependencies are phoning home.

**Comment by ashleymercer** [comment]
Also wishing to register my dislike.

As with OP I've been a very happy user of liquibase over the years and I'm enormously grateful to everyone who's contributed a pretty kick-ass tool to the world, for free! However also as with OP, I'm inclined to be dubious: I don't believe the maintainers' intentions are malicious but there are certainly several things that make me uncomfortable:

1. the [github release for 4.30.0](https://github.com/liquibase/liquibase/releases/tag/v4.30.0) only mentions analytics under the `[PRO]` section and not under the `[OSS]` section. As a user of the OSS version, I tend to mentally skip over the first part and assume it's not relevant. More fool me. Could this be updated / fixed please to make it VERY clear that this applies to all users?

2. As others commented, I don't believe the blog post is sufficient to raise awareness: I only came across this issue by accident when stuff started falling over. Would the maintainers accept a PR to add console output with a message explaining that analytics are now being collected, and pointing people the documentation? This sort of output has precedent e.g. previous addition of the liquibase banner, summary output for various commands and so on.

3. the wording on the [documentation page](https://docs.liquibase.com/analytics/home.html) is incredibly vague: many  items are phrased as "X data *like* a, b, c" - does that mean exactly a, b, and c, or does it mean other things as well? If you want people to trust this, can you please enumerate *exactly* what data is being collected? Ideally you would document every field in the JSON snippet above, explain exactly what each field means, why it needs to be collected, and how it will be used / aggregated.

4. following on from point 3: will the maintainers give a commitment that no additional fields will be collected in future, and that if they ever need to be, that will require a separate flag? Again, this is a trust issue: my concern is that some users actively (or more likely, passively) agree to the analytics as they are today, but that over time more and more stuff gets added to the analytics blob.

5. finally, can the maintainers please explain why analytics is disabled by default for paying customers? This seems illogical to me.

**Comment by lbadam** [comment]
@ashleymercer thanks for taking the time to write out your thoughts. We definitely understand the concern around analytics and usage tracking.

I'll do my best to address each one of your points:

1. I can update those release notes. There was no intention to mislead or cause users to gloss over anything. It was simply an issue of redundancy. I made the choice to not announce the same thing in two places. I can rewrite that and split up the announcement that is relevant to each group.

2. We initially explored a message like you describe, but decided to not do that for two reasons. First, Liquibase does not have a mechanism to show a banner only once (or any sort of dynamic handling). We have internally discussed the value of some sort of dynamic messaging system for various purposes (ie. updating to a new version). Second, with no way to limit the message like `homebrew` does to a first-time install or run, this meant that the message would be output every time a command is run. Liquibase output can already be verbose so we decided not to add that extra message to every action a user takes.

3. The body of the request can always be examined by using the global `--analytics-log-level=FINE` ([doc](https://docs.liquibase.com/parameters/analytics-log-level.html)). You will see every piece of data collected in the logs - again because we are not trying to obfuscate or hide what we are collecting. We have outlined in other places the kinds of insights we would like to extrapolate.

4. We do not currently have any intention of adding additional fields. I can't say we never will, but I can confidently say that we will not try to hide anything from our community and users. That means it would be visible in release notes, changelogs, and public messaging like blog posts, our Discord, and Forum.

5. I recently worked as a PM on another free, open-source CLI with opt-in analytics. Frankly, even though we had hundreds of thousands of downloads a week, no one ever opted-in. It severely limited our ability to understand how most users were using our product - limiting us to only the developers that wanted to talk to us. Interviews are great and necessary, and I love talking directly to users. That said, it is very limiting in terms of really understanding the breadth of user needs and desires. Liquibase is a project with a broad audience in terms of install methods and databases used. Implementing anonymous analytics that is opt-in by default for our free users and can be opted-out at any time with minimal effort is a much better solution for all involved.  For our paid users, we know them personally and already engage them directly and regularly about how they're using Liquibase. We encourage them to opt-in, but it's less of a need for understanding.

I hope that helps clear things up some. We do listen to all of the feedback, and don't take an addition like this lightly. We understand and value the privacy needs of our users.

**Comment by mydeveloperplanet** [comment]
@lbadam you do take the time to respond, that is very user friendly.

1. However, you mention the following:
_"I recently worked as a PM on another free, open-source CLI with opt-in analytics. Frankly, even though we had hundreds of thousands of downloads a week, no one ever opted-in."_
  And why would that be? Because **no one** want to opt-in. The solution is not to force them but to find other ways to reach your users.

2. _"and can be opted-out at any time with minimal effort "_
I strongly disagree. It costed us many days to disable this on all developer machines, images for new developer machines, CI-servers, all projects where Liquibase is used, customer sites, docker images, etc.

I really do hope that this opt-in is deleted as I am looking at Flyway now...

**Comment by npruehs** [comment]
Adding my two cents:

- I found this particularly frustrating because we pulled in analytics by updating Spring Boot, which transitively included a Liquibase update. I'm aware you're supposed to check release notes of third-party libraries included with any framework you're using, but that's definitely not something I've been looking out for.
- Even when disabling this, `liquibase.analytics.configuration.AnalyticsArgs` is logging the message "User has disabled analytics." at ERROR level. As a team that's spending a lot of time to keep our logs clean, and even has alerting set up for error logs, this is another thing I'd like to see changed in the future if possible.

Thanks for listening!

**Comment by golimpio** [comment]
I think instructions are clear and it's easy to disable analytics. But since I missed the announcement, this caused me some issues. It's my fault anyway, since I use Spring Boot for some services and I don't manage the Liquibase dependency manually, so I'm not very proactive in reading the announcements.

Though I don't agree with having analytics in libraries that can potentially be used in services running on the cloud, it all depends on your user base.

In my case, I have services running on AWS, and I have security in place to report and block outgoing access to external servers. When I first got the alerts, It wasn't clear that they were from Liquibase, I honestly wouldn't expect this from any dependency I have in my services.

This default behaviour was slowing down the start time for services and generating unnecessary alerts, and I imagine nowadays anyone concerned about security will run things in a similar way, blocking outgoing connections to unknown servers. Again, it's my fault for missing the announcement, and hopefully I was the only person to miss it, otherwise I can see the headache others will have if they use this library in a secure environment.

**Comment by golimpio** [comment]
One other item for thought: services running in secure environments where outgoing connections to external servers are blocked will never send analytics, so people will disable it or block it without Liquibase knowing about this, so I'm curious about how to measure this. I'd assume that we can't know how many users are turning it off or simply blocking it.

**Comment by ataraxus** [comment]
I would like to voice my concerns too. Aside from the GDPR topic. I was also caught by alerts from outgoing traffic. Since liquibase is also used by us via Spring, this was not very visible.
